### PR TITLE
HealDA layers and DiT patches for PNM Core

### PR DESCRIPTION
**Pull request by aayushg55**
# PhysicsNeMo Pull Request

## Description

Adds necessary layers and patches the existing DiT to enable HealDA integration into PhysicsNemo.

**DiT**
- Makes the DiT conditioning module for timestep and condition modular, allowing for custom `conditioning_embedder` modules. The existing embedding method embedded the timestep and condition separately and then added the two (following the original DiT implementation), as opposed to the EDM/SongUNet style approach, where the two are embedded jointly.
- Adds `DropPath` to the DiT
- Adds a `final_dropout` toggle to the `Mlp`
- Adds options for qk normalization in DiT attention with `timm` and TE (Note: TE does not support `qk_norm_affine` unlike ``timm``)

**HealDA**
- Scatter aggregation method for tokenizing sparse data onto a dense grid

**Breaking Changes/Bug fixes:**
- TE `MultiHeadAttention` defaults to sequence first `sbhd` qkv_format format matching PyTorch. This does not match the PNM documentation of `TESelfAttention`, suggesting batch-first `(B, L, D) `or the `timm` default of batch-first, so the `qkv_format` argument is exposed and the default is changed to `bshd` (batch-first) for consistency.
- TE  `MultiHeadAttention` does not support projection dropout, and so the appropriate dropout layer was skipped/not applied previously, making the timm and TE backends incompatible. Added a dropout layer following the `attn_op` for consistency. Breaking change for any model trained using TE backend with `proj_drop_rate > 0`.
- Previously in the DiTBlock, the dropout param in the Mlp was hardcoded to 0, so `mlp_drop_rate` was ignored. Changed to propagate this correctly. Breaking change to any model  trained with `mlp_drop_rate>0`
 

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: physicsnemo/experimental/models/dit/layers.py [hunk 1]
```diff
@@ -239,24 +268,46 @@ class TESelfAttention(AttentionModuleBase):
...
             warnings.warn(
-                "Transformer Engine MultiheadAttention does not support projection dropout (proj_drop_rate > 0). "
-                "The specified proj_drop_rate will be ignored."
+                "Transformer Engine does not support disabling affine parameters for QK norm. "
+                "Ignoring qk_norm_affine=False and using affine parameters.",
+                UserWarning,
+                stacklevel=2,
             )
-        self.attn_op = MultiheadAttention(hidden_size=hidden_size, num_attention_heads=num_heads, attention_dropout=attn_drop_rate, **kwargs)
+        kwargs.pop("qk_norm_affine", None)
```
[the remark this answers is not in the document]
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py:293]: TE's MultiheadAttention accepts qk_norm_type

**Comment by NickGeneva** [comment]
Moved HPX layers into module folder, can update this one now to use the embedding layers there

https://github.com/NVIDIA/physicsnemo/pull/1377

File: physicsnemo/experimental/models/dit/dit.py [hunk 2]
```diff
@@ -251,8 +261,11 @@ def __init__(
             self.detokenizer = detokenizer
 
 
+        # Linear drop_path schedule: 0 -> drop_path
+        drop_path_rates = [drop_path * i / max(1, depth - 1) for i in range(depth)]
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/dit.py]: In stead of hard coding the drop-path values to be this linear approach, perhaps its a bit more robust to make the entire parameter a list/array of floats and this gets replaced with some check that it is of the right length.

File: physicsnemo/experimental/models/dit/dit.py [hunk 3]
```diff
@@ -299,31 +314,20 @@ def forward(
         condition: Optional[torch.Tensor] = None,
         p_dropout: Optional[float | torch.Tensor] = None,
         attn_kwargs: Optional[Dict[str, Any]] = None,
+        tokenizer_kwargs: Optional[Dict[str, Any]] = None,
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/dit.py]: Can we just make all of these ` Dict[str, Any] = {}`, like why support None... maybe some legacy thing @pzharrington

File: physicsnemo/experimental/models/dit/dit.py
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/dit.py:59]: No a problem from you Aayush... but generally this whole doc string needs an update to the correct format others are in

File: physicsnemo/experimental/models/dit/layers.py [hunk 4]
```diff
@@ -239,24 +268,46 @@ class TESelfAttention(AttentionModuleBase):
...
+        **kwargs: Any,
+    ):
         super().__init__()
         if not TE_AVAILABLE:
             raise ImportError(
                 "Transformer Engine is not installed. Please install it with `pip install transformer-engine`."
             )
 
-        if proj_drop_rate > 0:
+        if "qk_norm_affine" in kwargs and not kwargs["qk_norm_affine"]:
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py:286]: Does the proj drop rate check still need to be here?

File: physicsnemo/experimental/models/dit/layers.py [hunk 5]
```diff
@@ -592,14 +651,14 @@ def forward(
             modulated_attn_input,
             **(attn_kwargs or {}),
         )
-        x = x + attention_gate.unsqueeze(1) * attention_output
+        x = torch.addcmul(x, self.drop_path(attention_gate.unsqueeze(1)), attention_output)
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py:659]: Will this impact the loading of previous DiTs (stromscope) since this is a unconditional module additon? @pzharrington

File: physicsnemo/experimental/models/dit/layers.py [hunk 6]
```diff
@@ -982,3 +1042,279 @@ def get_detokenizer(
             **detokenizer_kwargs,
         )
     raise ValueError("detokenizer must be 'proj_reshape_2d', no other supported detokenizers are available yet.")
+
+class ConditioningEmbedderBase(Module, ABC):
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py]: Can we move all these into a separate file.

Also can I suggest a protocol here instead of an abstract class... theres nothing getting inherited from `ConditioningEmbedderBase` so no need to force people to explicitly use this

File: physicsnemo/experimental/models/dit/layers.py [hunk 7]
```diff
@@ -982,3 +1042,279 @@ def get_detokenizer(
...
+                emb = emb + self.map_label(tmp * math.sqrt(self.map_label.in_features))
+
+        # MLP 
+        emb = torch.nn.functional.silu(self.map_layer0(emb))
+        emb = self.map_layer1(emb)
+
+        return emb
+
+
+def get_conditioning_embedder(
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py]: I would sugest moving this into the embedders file too

File: physicsnemo/experimental/models/dit/layers.py [hunk 8]
```diff
@@ -982,3 +1042,279 @@ def get_detokenizer(
...
+    hidden_size: int,
+    conditioning_embedder: Literal["dit", "edm", "zero"] = "dit",
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py]: Maybe an enum here is more appropriate where the values are the class objects of each?

File: physicsnemo/experimental/models/dit/layers.py [hunk 9]
```diff
@@ -982,3 +1042,279 @@ def get_detokenizer(
...
+    if conditioning_embedder == "dit":
+        return DiTConditionEmbedder(
+            hidden_size=hidden_size,
+            condition_dim=condition_dim,
+            **embedder_kwargs,
+        )
+    if conditioning_embedder == "edm":
+        return EDMConditionEmbedder(
+            emb_channels=condition_dim,
+            noise_channels=embedder_kwargs.pop("noise_channels", hidden_size),
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py]: whats up with the weird pop here....

shame all of these can have similar init signatures since we already pass `kwargs`

File: physicsnemo/experimental/models/dit/layers.py [hunk 10]
```diff
@@ -191,8 +199,26 @@ class TimmSelfAttention(AttentionModuleBase):
...
         Output tensor of shape (B, L, D).
     """
-    def __init__(self, hidden_size: int, num_heads: int, attn_drop_rate: float = 0.0, proj_drop_rate: float = 0.0, **kwargs: Any):
+    def __init__(
+        self,
+        hidden_size: int,
+        num_heads: int,
+        attn_drop_rate: float = 0.0,
+        proj_drop_rate: float = 0.0,
+        qk_norm_type: Optional[Literal["RMSNorm", "LayerNorm"]] = None,
```
Comment (NickGeneva) [inline physicsnemo/experimental/models/dit/layers.py]: Nit: Switch this to `Literal["RMSNorm", "LayerNorm"] | None`

File: physicsnemo/experimental/models/dit/dit.py (hunk 3 above)
Comment (pzharrington) [reply physicsnemo/experimental/models/dit/dit.py]: Yeah no need to support `None`, defaulting to `{}` sounds fine

File: physicsnemo/experimental/models/dit/layers.py (hunk 6 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py]: Moved the conditioning embedder classes into a separate file, using a protocol instead.

File: physicsnemo/experimental/models/dit/layers.py (hunk 5 above)
Comment (pzharrington) [reply physicsnemo/experimental/models/dit/layers.py:659]: Given that it's just a dropout module (no state dict params)  I think it should be fine. Tests should verify

**Comment by aayushg55** [comment]
Thanks @NickGeneva, updated to use the new hpx embedding layers.

File: physicsnemo/experimental/models/dit/layers.py (hunk 4 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py:286]: Previously, the `proj_drop_rate` was unused since TE does not natively support this dropout within its attention block. As we used this in HealDA, I added a dropout layer corresponding to this (and to match timm attention).
```
        # TE doesn't support proj_drop natively, so we add it manually after the attention output
        self.proj_drop = nn.Dropout(proj_drop_rate)
```

File: physicsnemo/experimental/models/dit/dit.py (hunk 2 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/dit.py]: Changed to take in a list of floats with length validation

File: physicsnemo/experimental/models/dit/dit.py (hunk 3 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/dit.py]: Changed the default to `{}`

File: physicsnemo/experimental/models/dit/layers.py (hunk 10 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py]: Changed

File: physicsnemo/experimental/models/dit/dit.py [hunk 11]
```diff
@@ -164,12 +177,14 @@ def __init__(
         attention_backend: Literal["timm", "transformer_engine", "natten2d"] = "timm",
         layernorm_backend: Literal["apex", "torch"] = "torch",
         condition_dim: Optional[int] = None,
+        conditioning_embedder: Union[Literal["dit", "edm"], Module] = "dit",
         dit_initialization: Optional[int] = True,
+        conditioning_embedder_kwargs: Dict[str, Any] = {},
         tokenizer_kwargs: Dict[str, Any] = {},
         detokenizer_kwargs: Dict[str, Any] = {},
         block_kwargs: Dict[str, Any] = {},
-        timestep_embed_kwargs: Dict[str, Any] = {},
```
Comment (pzharrington) [inline physicsnemo/experimental/models/dit/dit.py:171]: Removal of this will break the existing StormScope checkpoints, which pass `timestep_embed_kwargs`, specifically for the `np_sin_cos` setting of `embed_fn`. I'm not opposed to the more flexible `Protocol`-based approach in this PR for condition embedding, and the more sensible name `conditioning_embedder_kwargs`, but we should use the backward compat checkpoint handling to make sure those legacy checkpoints can be loaded without error. @aayushg55 I can do that part in a follow-up PR if you don't want to bother with it. However can you update `DiTConditionEmbedder` to properly accept and route all the kwarg options for `PositionalEmbedding`?

File: physicsnemo/experimental/models/dit/layers.py (hunk 6 above)
Comment (NickGeneva) [reply physicsnemo/experimental/models/dit/layers.py]: awesome!

File: physicsnemo/experimental/models/dit/layers.py (hunk 5 above)
Comment (NickGeneva) [reply physicsnemo/experimental/models/dit/layers.py:659]: Will leave this unresolved, until we confirm stormscope doesnt break... what would be the best way to verify that checkpoint runs fine?

something like, with the [hf files](https://huggingface.co/nvidia/stormscope-goes-mrms)?
```python
registry = json.load("registry.json")
pkg = registry[model_name]

model_spec = []
for m in pkg["checkpoints"]:
  model = Module.from_checkpoint(m["path"])
  model_spec.append(
      {
          "model": model_wrap(model),
          "sigma_min": float(m["sigma_min"]),
          "sigma_max": float(m["sigma_max"]),
      }
  )
```

e2s doesnt support psn 2.x yet so it will break

File: physicsnemo/experimental/models/dit/dit.py (hunk 3 above)
Comment (NickGeneva) [reply physicsnemo/experimental/models/dit/dit.py]: thanks!

File: physicsnemo/experimental/models/dit/layers.py (hunk 4 above)
Comment (NickGeneva) [reply physicsnemo/experimental/models/dit/layers.py:286]: Ah I see. Thanks for clarifying

File: physicsnemo/experimental/models/dit/layers.py (hunk 8 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py]: Changed to use enums:
```
class ConditioningEmbedderType(Enum):
    """Conditioning embedder types for DiT models."""

    DIT = DiTConditionEmbedder
    EDM = EDMConditionEmbedder
    ZERO = ZeroConditioningEmbedder

```

File: physicsnemo/experimental/models/dit/layers.py (hunk 9 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/layers.py]: Cleaned up the initialization to not require popping and instead just propagate the `kwargs`

File: physicsnemo/experimental/models/dit/dit.py (hunk 11 above)
Comment (aayushg55) [reply physicsnemo/experimental/models/dit/dit.py:171]: I updated it to propagate all the kwargs to `PositionalEmbedding` as before. Would be great if you could help with the backward compatibility checkpoint handling.

**Comment by NickGeneva** [comment]
Over all I think this looks good, only big pending item left  imois verifying that stormscope checkpoints remain operational. Other than than the changes reasonable extensions that dont have major api breaks.

File: physicsnemo/experimental/models/dit/dit.py (hunk 11 above)
Comment (pzharrington) [reply physicsnemo/experimental/models/dit/dit.py:171]: I added some checkpoint compat handling here: https://github.com/pzharrington/physicsnemo/commit/ac42b55d330d2d1df6b19b8c10134ec14f363a8e

Will try and test it tomorrow in E2S, if it works then I'll just push it to this branch/PR. @NickGeneva for viz

Comment (pzharrington) [reply physicsnemo/experimental/models/dit/dit.py:171]: Pushed the updates to this branch, test of the stormscope model with it looks good:
image

File: physicsnemo/experimental/models/dit/dit.py
Comment (pzharrington) [reply physicsnemo/experimental/models/dit/dit.py:59]: Will address this in un-migrating DiT from experimental

File: physicsnemo/nn/module/mlp_layers.py [hunk 12]
```diff
@@ -95,6 +95,7 @@ def __init__(
         out_features: int | None = None,
         act_layer: nn.Module | type[nn.Module] | str = nn.GELU,
         drop: float = 0.0,
+        final_dropout: bool = True,
```
Comment (pzharrington) [inline physicsnemo/nn/module/mlp_layers.py:92]: @aayushg55 can you add description of this to the docstring?

Comment (aayushg55) [reply physicsnemo/nn/module/mlp_layers.py:92]: Just added
